### PR TITLE
feature: implement razer pro click support

### LINKED ISSUE #1389: Support for Pro Click
### Device Info

* Name: Pro Click
* Product Number: RZ01-02990100
* Model Number: RZ01-0299

### Device Information

Please post the output of
```
    lsusb -d 1532: -v

Bus 001 Device 028: ID 1532:0077 Razer USA, Ltd Razer Pro Click
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0
  bDeviceSubClass         0
  bDeviceProtocol         0
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x0077
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer Pro Click
  iSerial                 0
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0054
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      1 Keyboard
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     159
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      1 Keyboard
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
```

```
Please post the output of
    ls -lah /dev/input/by-id/

total 0
drwxr-xr-x 2 root root 160 гру 25 15:49 .
drwxr-xr-x 4 root root 700 гру 25 15:49 ..
lrwxrwxrwx 1 root root  10 гру 25 15:45 usb-Razer_Razer_Pro_Click-event-if01 -> ../event13
lrwxrwxrwx 1 root root  10 гру 25 15:45 usb-Razer_Razer_Pro_Click-event-mouse -> ../event11
lrwxrwxrwx 1 root root  10 гру 25 15:45 usb-Razer_Razer_Pro_Click-if01-event-kbd -> ../event12
lrwxrwxrwx 1 root root  10 гру 25 15:45 usb-Razer_Razer_Pro_Click-if02-event-kbd -> ../event16
lrwxrwxrwx 1 root root   9 гру 25 15:45 usb-Razer_Razer_Pro_Click-mouse -> ../mouse0
lrwxrwxrwx 1 root root  10 гру 25 14:21 usb-SteelSeries_SteelSeries_KLC-event-if01 -> ../event19
```

### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### PR DESCRIPTION
**Pull request by lukassup**
Implements some support for Razer Pro Click Wired & Receiver. Bluetooth does not create an event device file so maybe we can't control it over Bluetooth.

So far in `polychromatic` I'm able to control DPI (incl DPI stages), idle time, polling rate, low battery threshold.

Not sure which `razer_get_report` wait times to choose for the driver.

Let me know if you would like to make any amends.

Should close #1389

### CODE REVIEW

**Comment by lukassup** [comment]
## Receiver
image

**Comment by lukassup** [comment]
## Wired
image

**Comment by lukassup** [comment]
No event device is created in `/dev/input/by-id` when connected over Bluetooth so I guess I can't control it

File: driver/razermouse_driver.h [hunk 1]
```diff
@@ -108,7 +111,7 @@ struct razer_mouse_device {
     unsigned int tilt_repeat;
     __s32 hwheel_value;
     u8 button_byte; // Previous value of mouse button byte in HID record
-    u8 rep4[16]; // Previous value of report 4 on the keyboard intf
+    u8 rep4[16];    // Previous value of report 4 on the keyboard intf
```
Comment (z3ntu) [inline driver/razermouse_driver.h]: remove this formatting change

File: driver/razermouse_driver.h [hunk 2]
```diff
@@ -79,9 +82,9 @@
 /* Each keyboard report has 90 bytes*/
 #define RAZER_REPORT_LEN 0x5A
 
-#define RAZER_MAMBA_ROW_LEN 15          // 0 => 14
-#define RAZER_MAMBA_TE_ROW_LEN 16       // 0 => 15
-#define RAZER_DIAMONDBACK_ROW_LEN 21    // 0 => 20
+#define RAZER_MAMBA_ROW_LEN 15       // 0 => 14
+#define RAZER_MAMBA_TE_ROW_LEN 16    // 0 => 15
+#define RAZER_DIAMONDBACK_ROW_LEN 21 // 0 => 20
```
Comment (z3ntu) [inline driver/razermouse_driver.h]: remove these formatting changes

**Review by z3ntu — APPROVED** [review]
Apart from the minor nitpicks, looks pretty good! I'll let this still sit for a bit so maybe others get a chance to test this.

**Comment by lukassup** [comment]
I guess I'll also remove Bluetooth related code as it does not appear to work

**Review by z3ntu — APPROVED** [review]
LGTM, thanks again!

**Pull request merged**